### PR TITLE
Fix missing ISerializable interface on some types in System.Drawing

### LINKED ISSUE #22599: System.Drawing types lost ISerializable
This is showing up in /src/shims/ApiCompatBaseline.netcoreapp.netfx461.txt

```
Compat issues with assembly System.Drawing:
CannotRemoveBaseTypeOrInterface : Type 'System.Drawing.Bitmap' does not implement interface 'System.Runtime.Serialization.ISerializable' in the implementation but it does in the contract.
CannotMakeTypeAbstract : Type 'System.Drawing.BufferedGraphicsManager' is abstract in the implementation but is not abstract in the contract.
CannotRemoveBaseTypeOrInterface : Type 'System.Drawing.Font' does not implement interface 'System.Runtime.Serialization.ISerializable' in the implementation but it does in the contract.
CannotRemoveBaseTypeOrInterface : Type 'System.Drawing.Icon' does not implement interface 'System.Runtime.Serialization.ISerializable' in the implementation but it does in the contract.
CannotRemoveBaseTypeOrInterface : Type 'System.Drawing.Image' does not implement interface 'System.Runtime.Serialization.ISerializable' in the implementation but it does in the contract.
CannotRemoveBaseTypeOrInterface : Type 'System.Drawing.Imaging.Metafile' does not implement interface 'System.Runtime.Serialization.ISerializable' in the implementation but it does in the contract.
MembersMustExist : Member 'System.Drawing.Printing.InvalidPrinterException..ctor(System.Runtime.Serialization.SerializationInfo, System.Runtime.Serialization.StreamingContext)' does not exist in the implementation but it does exist in the contract.
```

For types we don't support binary serializing, like these, we are implementing the interface but throwing PNSE. This at least allows casting to work. Eg.,

```c#
            private Enumerator(SerializationInfo info, StreamingContext context)
            {
                throw new PlatformNotSupportedException();
            }

            void ISerializable.GetObjectData(SerializationInfo info, StreamingContext context)
            {
                throw new PlatformNotSupportedException();
            }
```

We should do thes ame for drawing and then msbuild src/shims /p:baselineapicompat=true to remove the errors.

### PR DESCRIPTION
**Pull request by DonaldHarris**
Fix missing ISerializable interface on some types in System.Drawing. Fixes #22599.

### CODE REVIEW

File: src/System.Drawing.Common/src/System/Drawing/Bitmap.NotSerializable.cs [hunk 1]
```diff
@@ -0,0 +1,16 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
```
Comment (ViktorHofer) [inline src/System.Drawing.Common/src/System/Drawing/Bitmap.NotSerializable.cs]: Bitmap itself isn't declared ISerializable: http://referencesource.microsoft.com/#System.Drawing/commonui/System/Drawing/Bitmap.cs,37

Its parent type `System.Drawing.Image` is. You should remove this file

File: src/System.Drawing.Common/src/System/Drawing/Bitmap.cs [hunk 2]
```diff
@@ -8,11 +8,12 @@
 using System.IO;
 using System.Runtime.InteropServices;
 using System.Security.Permissions;
+using System.Runtime.Serialization;
 
 namespace System.Drawing
 {
     [ComVisible(true)]
-    public sealed partial class Bitmap : Image
+    public sealed partial class Bitmap : Image, ISerializable
```
Comment (ViktorHofer) [inline src/System.Drawing.Common/src/System/Drawing/Bitmap.cs]: See comment above. Bitmap isn't implementing ISerializable by itself but by `System.Drawing.Image`. Remove it here.

File: src/System.Drawing.Common/src/System.Drawing.Common.csproj [hunk 3]
```diff
@@ -181,6 +181,7 @@
     <Compile Include="AssemblyRef.cs" />
     <Compile Include="SRDescriptionAttribute.cs" />
     <Compile Include="System\Drawing\Bitmap.cs" />
+    <Compile Include="System\Drawing\Bitmap.NotSerializable.cs" />
```
Comment (ViktorHofer) [inline src/System.Drawing.Common/src/System.Drawing.Common.csproj]: See comment above again. Remove this line.

File: src/System.Drawing.Common/src/System/Drawing/Imaging/Metafile.NotSerializable.cs [hunk 4]
```diff
@@ -0,0 +1,16 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
```
Comment (ViktorHofer) [inline src/System.Drawing.Common/src/System/Drawing/Imaging/Metafile.NotSerializable.cs]: Same as Bitmap. Metafile isn't implementing ISerializable directly: http://referencesource.microsoft.com/#System.Drawing/commonui/System/Drawing/Advanced/Metafile.cs,36. Remove this file.

File: src/System.Drawing.Common/src/System/Drawing/Imaging/Metafile.cs [hunk 5]
```diff
@@ -5,14 +5,15 @@
...
+using System.Runtime.Serialization;
 
 namespace System.Drawing.Imaging
 {
     /// <summary>
     /// Defines a graphic metafile. A metafile contains records that describe a sequence of graphics operations that
     /// can be recorded and played back.
     /// </summary>
-    public sealed partial class Metafile : Image
+    public sealed partial class Metafile : Image, ISerializable
```
Comment (ViktorHofer) [inline src/System.Drawing.Common/src/System/Drawing/Imaging/Metafile.cs]: Remove ISerializable here, Image is already implementing ISerializable

File: src/System.Drawing.Common/src/System.Drawing.Common.csproj [hunk 6]
```diff
@@ -195,17 +196,21 @@
...
     <Compile Include="System\Drawing\Icon.cs" />
+    <Compile Include="System\Drawing\Icon.NotSerializable.cs" />
     <Compile Include="System\Drawing\Image.cs" />
+    <Compile Include="System\Drawing\Image.NotSerializable.cs" />
     <Compile Include="System\Drawing\ImageAnimator.cs" />
     <Compile Include="System\Drawing\ImageInfo.cs" />
     <Compile Include="System\Drawing\Imaging\BitmapData.cs" />
     <Compile Include="System\Drawing\Imaging\ImageFormat.cs" />
     <Compile Include="System\Drawing\Imaging\Metafile.cs" />
+    <Compile Include="System\Drawing\Imaging\Metafile.NotSerializable.cs" />
```
Comment (ViktorHofer) [inline src/System.Drawing.Common/src/System.Drawing.Common.csproj]: Remove this file.

**Comment by ViktorHofer** [comment]
CannotMakeTypeAbstract : Type 'System.Drawing.BufferedGraphicsManager' is abstract in the implementation but is not abstract in the contract.
Netfx: http://referencesource.microsoft.com/#System.Drawing/commonui/System/Drawing/BufferedGraphicsManager.cs,8697d8f126ceea57
Core: https://github.com/dotnet/corefx/blob/master/src/System.Drawing.Common/src/System/Drawing/BufferedGraphicsManager.cs

@mellinoe I guess that change is intentional? For intentional changes we put these error messages into the ApiCompatBaseline.ignore file with a comment so that it won't pop up again. If it's intentional we can ask @DonaldHarris to do it in this PR.

**Comment by akoeplinger** [comment]
@ViktorHofer yes that was intentional, there were a couple of sealed classes with non-public constructors (which basically makes them like static classes) in System.Drawing so it was decided to just change them, some of them were done here: https://github.com/dotnet/corefx/pull/21293

**Comment by ViktorHofer** [comment]
thanks @akoeplinger. I updated the baselines.

**Comment by DonaldHarris** [comment]
@ViktorHofer I fixed the things you said to, but when I tried the command for the baseline file, it didn't change any of the System.Drawing lines. What do I do to make that work?

**Comment by ViktorHofer** [comment]
I guess you haven't built the Sustem.Drawing.Common.dll before exciting the command? If it's still not working don't bother with it

File: src/System.Drawing.Common/src/System/Drawing/Printing/InvalidPrinterException.cs [hunk 7]
```diff
@@ -22,6 +23,11 @@ public InvalidPrinterException(PrinterSettings settings)
             _settings = settings;
         }
 
+        protected InvalidPrinterException(SerializationInfo info, StreamingContext context) : base(info, context)
```
Comment (ViktorHofer) [inline src/System.Drawing.Common/src/System/Drawing/Printing/InvalidPrinterException.cs]: If Mono uses this file InvalidPrinterException.Serializable.cs together they will have a duplicate method signature and the compiler will emit an error. I would put that method into a NonSerializable file like you did for the others.

**Comment by DonaldHarris** [comment]
@ViktorHofer , I tried building System.Drawing.Common.dll, but it didn't help. I fixed the InvalidPrinterException constructor though.

**Comment by DonaldHarris** [comment]
@ViktorHofer I don't think I broke those tests. Is it good enough to merge?

**Comment by ViktorHofer** [comment]
@dotnet-bot test Linux x64 Release Build
@dotnet-bot test Windows x86 Release Build

**Comment by ViktorHofer** [comment]
I updated the ref code for you.

**Comment by ViktorHofer** [comment]
@mellinoe please review

**Comment by ViktorHofer** [comment]
Aah Eric is OOF until 8/24.

@karelz should we wait or let somebody else review?

**Comment by DonaldHarris** [comment]
@ViktorHofer ok thanks for fixing the ref file.

File: src/System.Drawing.Common/src/System/Drawing/Imaging/Metafile.cs [hunk 8]
```diff
@@ -5,6 +5,7 @@
 using System.Runtime.InteropServices;
 using System.IO;
 using System.Drawing.Internal;
+using System.Runtime.Serialization;
```
Comment (danmosemsft) [inline src/System.Drawing.Common/src/System/Drawing/Imaging/Metafile.cs]: revert?

File: src/System.Drawing.Common/src/System/Drawing/Bitmap.cs [hunk 9]
```diff
@@ -8,6 +8,7 @@
 using System.IO;
 using System.Runtime.InteropServices;
 using System.Security.Permissions;
+using System.Runtime.Serialization;
```
Comment (danmosemsft) [inline src/System.Drawing.Common/src/System/Drawing/Bitmap.cs]: revert

File: src/System.Drawing.Common/src/System/Drawing/Printing/InvalidPrinterException.NotSerializable.cs [hunk 10]
```diff
@@ -0,0 +1,15 @@
...
+// See the LICENSE file in the project root for more information.
+
+using System.Runtime.Serialization;
+
+namespace System.Drawing.Printing
+{
+    partial class InvalidPrinterException
+    {
+        protected InvalidPrinterException(SerializationInfo info, StreamingContext context) : base(info, context)
+        {
```
Comment (danmosemsft) [inline src/System.Drawing.Common/src/System/Drawing/Printing/InvalidPrinterException.NotSerializable.cs]: throw new PlatformNotSupportedException();

**Comment by danmosemsft** [comment]
@ViktorHofer I don't think this needs to wait on @mellinoe

**Comment by danmosemsft** [comment]
Thanks @DonaldHarris for the contribution. Want to grab something else now?

**Comment by ViktorHofer** [comment]
@DonaldHarris thanks for the contribution! 👍 Happy to have you around.

**Pull request merged**